### DISCUSSION TITLE
OpenModelica CMake build instructions for MSVC encountered an error

### DISCUSSION THREAD

**Discussion by pily1** [OpenModelica/OpenModelica#12633] [Q&A] [answered]
hi,I followed the https://github.com/OpenModelica/OpenModelica/blob/master/README.cmake.msvc.md file,but after the 5 step:Configure OpenModelica,it encountered an error ,Make Error at OMCompiler/Compiler/CMakeLists.txt:299 (message):
  Environment variable "MSYSTEM_PREFIX" is not set.
the following image is the process
![image](https://github.com/OpenModelica/OpenModelica/assets/20216269/f0666e39-4c6c-4c82-98af-1d9abd77f2d6)

**Comment by AnHeuermann** [comment]
The environment variable `MSYSTEM_PREFIX` shouldn't be used for the MSVC variant of the Windows build. But we don't test it, so that's where this error is coming from.

When using the MSYS2 shell provided by OMDev and generator `-G "MSYS Makefiles"` the environment variable is defined.
See https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/README.Windows.md#21-msys-and-cmake for the details.

How to solve this for the ninja build? We need to differentiate between MINGW and MSVC case in the CMakeListst.txt. At [OMCompiler/Compiler/CMakeLists.txt:299](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/Compiler/CMakeLists.txt#L292-L318) we need an additional check for `MINGW` and get the `DIRECTORIES` of the runtime dependencies. 
For now you can disable it and see if that solves your issue.

```diff
# Install OpenModelicaCompiler to the directory for LIBRARIES (default=lib)
# Install omc to the directory for RUNTIME (default=bin)
# They are part of "compiler" installation component. This means that you can
# install just the files in this component by specifying its name.
-if(WIN32)
+if(WIN32 AND MINGW)
  # Escape the environment variable path
  if(NOT DEFINED ENV{MSYSTEM_PREFIX})
    message(FATAL_ERROR "Environment variable \"MSYSTEM_PREFIX\" is not set.")
  endif()
  string(REPLACE "\\" "/" MSYSTEM_PREFIX_ESCAPED "$ENV{MSYSTEM_PREFIX}")

  set(OMCOMPILER_LIB_DIR ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}OMCompiler/Compiler/bin)
  set(SIMULATION_RUNTIME_LIB_DIR ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}OMCompiler/SimulationRuntime/c)
  set(GC_RUNTIME_LIB_DIR ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}OMCompiler/3rdParty/gc)
  set(RUNTIME_LIB_DIRS ${OMPLOT_LIB_DIR} ${OMSIMULATOR_LIB_DIR} ${OMCOMPILER_LIB_DIR} ${SIMULATION_RUNTIME_LIB_DIR} ${GC_RUNTIME_LIB_DIR})

  install(TARGETS OpenModelicaCompiler
          RUNTIME_DEPENDENCIES
            DIRECTORIES ${MSYSTEM_PREFIX_ESCAPED}/bin ${RUNTIME_LIB_DIRS}
            PRE_EXCLUDE_REGEXES "api-ms-" "ext-ms-"
            POST_EXCLUDE_REGEXES ".*system32/.*\\.dll")
  install(TARGETS omc
          COMPONENT omc)
else(WIN32)
  install(TARGETS omc OpenModelicaCompiler
          COMPONENT omc)
endif(WIN32)
```

You can change the CMakeLists.txt in the SimulationRuntime/cpp in a similar fashion:

```diff
IF(WIN32)
	SET(Boost_LIBS_ "$(OMHOME)/bin")
-	if(NOT DEFINED ENV{MSYSTEM_PREFIX})
-		message(FATAL_ERROR "Environment variable \"MSYSTEM_PREFIX\" is not set.")
-	endif()
-	string(REPLACE "\\" "/" MSYSTEM_PREFIX_ESCAPED "$ENV{MSYSTEM_PREFIX}")
	IF(MSVC)
		SET(Boost_INCLUDE_ "$(OMHOME)/include/omc/omsicpp/3rdParty/boost/msvc/")
	ELSE(MSVC)
+		if(NOT DEFINED ENV{MSYSTEM_PREFIX})
+			message(FATAL_ERROR "Environment variable \"MSYSTEM_PREFIX\" is not set.")
+		endif()
+		string(REPLACE "\\" "/" MSYSTEM_PREFIX_ESCAPED "$ENV{MSYSTEM_PREFIX}")
		SET(Boost_INCLUDE_ "${MSYSTEM_PREFIX_ESCAPED}/include/")
	ENDIF(MSVC)
```

**Reply by AnHeuermann** [reply]
I was able to configure with CMake and ninja, but building with Visual Studio 2019 failed. I don't think the MSVC build is still operational.

**Reply by pily1** [reply]
I modified CMakeLists.txt according to your instructions，the second CMakeLists.txt is in the SimulationRuntime/cpp/Core/Modelica,
I used the "Developer PowerShell for VS 2022" terminal,after executing the command "cmake --build build_msvc_ninja --target install",it encountered errors,"[4/2731] Linking CXX shared library OMCompiler\SimulationRuntime\cpp\Solver\OMCppCVode.dll
FAILED: OMCompiler/SimulationRuntime/cpp/Solver/OMCppCVode.dll OMCompiler/SimulationRuntime/cpp/Solver/OMCppCVode.lib"
the following image is the process:
![image](https://github.com/OpenModelica/OpenModelica/assets/20216269/d648b229-a021-4022-923f-821862d2f307)

**Reply by AnHeuermann** [reply]
Do you need the MSVC version of the build or would the usual UCRT64 version work for you as well (the one you'll get if you install OpenModelica)?

**Reply by pily1** [reply]
no,I want to debug the source code with Visual Studio 2022 or Visual Studio Code,so I want to compile a debug version by myself。

**Reply by AnHeuermann** [reply]
You can build with MSYS (UCRT64) and debug with gdb. This usually works, unless you want to debug something like an external function from a Modelica library that was compiled with MSVC.

**Reply by pily1** [reply]
I have tried it,but when I debug with gdb,it enters the C code, the watched value is void* type,it's a address,I can't get the original value,so I can't understand the input and output param。

**Reply by AnHeuermann** [reply, marked as answer]
That won't change with MSVC. The issue is that omc use some boxed memory pointers. The memory layout can be found in https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/SimulationRuntime/c/meta/meta_modelica.c
You would need to cast them manually or call [`printAny`](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/SimulationRuntime/c/meta/meta_modelica.c#L435) on the void pointers. It's a nightmare to debug and I usually give up and add prints everywhere. I started creating a VS Code debugger, but I probably won't finish it.

**Reply by pily1** [reply]
think you very much,I will try it。
